### ISSUE TITLE
Specialized methods on generic classes

### ISSUE THREAD

**Issue by Tinche** [python/mypy#14243]
**Bug Report**

Let's say I have a generic class `G`.

```python
from __future__ import annotations

from collections.abc import Sized
from typing import Generic, TypeVar

T = TypeVar("T")

class G(Generic[T]):
    pass
```

I want to have a single method on this class that's only legal for a subset of `T`.

If I want to limit it to, say, `int`, it works:

```python
class G(Generic[T]):
    def test(self: G[int]) -> G[int]:
        return self

G[int]().test()  # Works, as it should
G[str]().test()  # Error, as it should
```

But if I want to limit it to, say, a protocol like `Sized`, I get a false negative:

```python
S = TypeVar("S", bound=Sized)

class G(Generic[T]):
    def test(self: G[S]) -> G[S]:
        return self


G[int]().test()  # False negative - works, but it shouldn't
```

This is a pity, since it'd unlock some cool features I was working on vis-a-vis using attrs attributes. Are there any workaround? Would this be handled by intersection types maybe?

**Your Environment**

- Mypy version used: mypy 1.0.0+dev.b8c03ab6809aab56928f3cd865edb44944a600a2
- Mypy command-line flags:
- Mypy configuration options from `mypy.ini` (and other config files):
- Python version used: Python 3.10.6

**Comment by CaselIT** [comment]
Another example of this bug is the following
https://mypy-play.net/?mypy=latest&python=3.14&gist=2b566dc3fa9787cf75b7517168ad2ff7
```py
from typing import Any, Generic, TypeVar, assert_type, overload
class A:
    pass

class C(A):
    pass

_a = TypeVar("_a", bound=A)
_a1 = TypeVar("_a1", bound=A)
_c = TypeVar("_c", bound=C)

class G(Generic[_a]):
    x: _a

    @overload
    def go(self: "G[_c]") -> int: ...
    @overload  # using _a is the same
    def go(self: "G[_a1]") -> str | int: ...

    def go(self) -> int | str:
        return 1 if isinstance(self.x, C) else "a"

    @overload
    def og(self, a: "G[_c]") -> int: ...
    @overload  # using _a is the same
    def og(self, a: "G[_a1]") -> str | int: ...

    def og(self, a: "G[Any]") -> int | str:
        return 1 if isinstance(self.x, C) else "a"

def a(x: G[A], y: G[C]) -> None:
    assert_type(x.go(), str | int)
    assert_type(y.go(), int)
    assert_type(x.og(x), str | int)
    assert_type(y.og(y), int)
```

this should not error but instead  mypy thinks G[A] is G[C] if the constraint is on self, works fine if it's an argument instead

> main.py:37: error: Expression is of type "int", not "str | int"  [assert-type]
